go/libraries/doltcore/sqle/dolt_schema_diff_table_function.go

```diff
@@ -0,0 +1,420 @@
+	&sql.Column{Name: "to_create_statement", Type: types.Text, Nullable: false},   // 3
+	&sql.Column{Name: "alter_statement", Type: types.Text, Nullable: false},       // 4
+}
+
+// NewInstance creates a new instance of TableFunction interface
+func (ds *SchemaDiffTableFunction) NewInstance(ctx *sql.Context, db sql.Database, expressions []sql.Expression) (sql.Node, error) {
+	newInstance := &SchemaDiffTableFunction{
+		ctx:      ctx,
+		database: db,
+	}
+
+	node, err := newInstance.WithExpressions(expressions...)
+	if err != nil {
+		return nil, err
+	}
+
+	return node, nil
+}
+
+// Database implements the sql.Databaser interface
+func (ds *SchemaDiffTableFunction) Database() sql.Database {
+	return ds.database
+}
+
+// WithDatabase implements the sql.Databaser interface
+func (ds *SchemaDiffTableFunction) WithDatabase(database sql.Database) (sql.Node, error) {
+	nds := *ds
+	nds.database = database
+	return &nds, nil
+}
+
+// Name implements the sql.TableFunction interface
+func (ds *SchemaDiffTableFunction) Name() string {
+	return "dolt_schema_diff"
+}
+
+func (ds *SchemaDiffTableFunction) commitsResolved() bool {
+	if ds.dotCommitExpr != nil {
+		return ds.dotCommitExpr.Resolved()
+	}
+	return ds.fromCommitExpr.Resolved() && ds.toCommitExpr.Resolved()
+}
+
+// Resolved implements the sql.Resolvable interface
+func (ds *SchemaDiffTableFunction) Resolved() bool {
+	if ds.tableNameExpr != nil {
+		return ds.commitsResolved() && ds.tableNameExpr.Resolved()
+	}
+	return ds.commitsResolved()
+}
+
+// String implements the Stringer interface
+func (ds *SchemaDiffTableFunction) String() string {
+	if ds.dotCommitExpr != nil {
+		if ds.tableNameExpr != nil {
+			return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s)", ds.dotCommitExpr.String(), ds.tableNameExpr.String())
+		} else {
+			return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s)", ds.dotCommitExpr.String())
+		}
+	}
+	if ds.tableNameExpr != nil {
+		return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s, %s)", ds.fromCommitExpr.String(), ds.toCommitExpr.String(), ds.tableNameExpr.String())
+	} else {
+		return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s)", ds.fromCommitExpr.String(), ds.toCommitExpr.String())
+	}
+}
+
+// Schema implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) Schema() sql.Schema {
+	return schemaDiffTableSchema
+}
+
+// Children implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) Children() []sql.Node {
+	return nil
+}
+
+// WithChildren implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 0 {
+		return nil, fmt.Errorf("unexpected children")
+	}
+	return ds, nil
+}
+
+// CheckPrivileges implements the interface sql.Node.
+func (ds *SchemaDiffTableFunction) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	if ds.tableNameExpr == nil || !types.IsText(ds.tableNameExpr.Type()) {
```

Comment:
Why does the table name not being text mean that the user has privileges? This might warrant a comment.

Reply:
Good point. In other cases, if the text isn't a string, we return `false`. Updating.

Reply:
My question is less "what should we do here?" but "why do we do what we do?" It's not clear to me why we should return any particular value in this case. Is it because we expect the tableName to be text but the function doesn't allow us to return an error?

Reply:
Very good question, but I don't have an answer yet.
We are definitely checking the type in `WithExpressions`, so in theory there should be no way for a wrong type of expression to make it into this command. But I'm not, and all of our other implementations of `CheckPrivileges` return `false` when the expressions are of the wrong type, and all of those functions have those same checks as well.
@fulghum , do you know?

Reply:
Generally, when I see this in a `CheckPrivileges` function, it's because the user could pass in other expressions that don't have a string type (technically we would/should coerce anything to a string type, but we're being a little restrictive here since we know we need a valid string value that isn't coerced from another type) so before we cast to a string, we do this check and fail the privilege check if it's not a string. This is a little bit of extra defense, since validation errors in other places should catch this before the plan is going to be executed and starts checking privileges. In this case, it doesn't look like we ever cast the value to a string, so we probably don't need this extra check here.